### PR TITLE
[BIONEMO-2473] Added tests for Evo2 LoRA fine-tuning

### PR DESCRIPTION
**Pull request by balvisio**
### Description

Fixes and added test for Evo2LoRA

### Type of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
# TODO: Add code snippet
```

### Pre-submit Checklist

- [x] I have tested these changes locally
- [x] I have updated the documentation accordingly
- [x] I have added/updated tests as needed
- [ ] All existing tests pass successfully

### CODE REVIEW

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py [hunk 1]
```diff
@@ -30,14 +30,14 @@
 from bionemo.testing.megatron_parallel_state_utils import distributed_model_parallel_state
 from bionemo.testing.subprocess_utils import run_command_in_subprocess
 
+from .common import small_training_cmd, small_training_finetune_cmd
```
Comment (yzhang123) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py:33]: could you pls use full path starting with bionemo...

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_lora.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import pytest
+
+from bionemo.testing.subprocess_utils import run_command_in_subprocess
+
+from .common import small_training_cmd, small_training_finetune_cmd
```
Comment (yzhang123) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_lora.py:25]: pls use conventional import path bionemo....

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py (hunk 1 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py:33]: Unfortunately, the test environment doesn't support absolute imports because of the packaging. I had to keep package-relative imports.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_lora.py (hunk 2 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_lora.py:25]: Unfortunately, the test environment doesn't support absolute imports because of the packaging. I had to keep package-relative imports.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_lora.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import pytest
+
+from bionemo.evo2.run.common import small_training_cmd, small_training_finetune_cmd
+from bionemo.testing.subprocess_utils import run_command_in_subprocess
+
+
+@pytest.mark.timeout(512)  # Optional: fail if the test takes too long.
+@pytest.mark.slow
+@pytest.mark.parametrize("with_peft", [True, False])
+def test_train_evo2_finetune_runs_lora(tmp_path, with_peft: bool):
```
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_lora.py]: @balvisio , could you test in this test if the reduced train loss and val loss decreases?

Comment (dorotat-nv) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_lora.py]: (series can be takes from tensorboard event files)

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+# limitations under the License.
+import pytest
+
+from bionemo.evo2.run.common import small_training_cmd, small_training_finetune_cmd
+from bionemo.testing.subprocess_utils import run_command_in_subprocess
+
+
+@pytest.mark.timeout(512)  # Optional: fail if the test takes too long.
+@pytest.mark.slow
+@pytest.mark.parametrize("with_peft", [True, False])
```
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py:53]: does this param controls if lora is used? if yes, then lets rename the test to test_train_evo2_finetune_runs

Comment (dorotat-nv) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py:53]: and this test file to test_finetune

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/common.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        f"train_evo2 --mock-data --result-dir {path} --devices {devices} "
+        "--model-size 1b_nv --num-layers 4 --hybrid-override-pattern SDH* --limit-val-batches 1 "
+        "--no-activation-checkpointing --add-bias-output --create-tensorboard-logger --create-tflops-callback "
+        f"--max-steps {max_steps} --warmup-steps 1 --val-check-interval {val_check} --limit-val-batches 1 "
+        f"--seq-length 16 --hidden-dropout 0.1 --attention-dropout 0.1 {additional_args}"
+    )
+    return cmd
+
+
+def small_training_finetune_cmd(
```
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/common.py:40]: since it is only used in finetune, lets keep this method as in test_finetune

Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/common.py:40]: `small_training_finetune_cmd` is used both in `test_lora.py` and `test_train.py`

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py (hunk 4 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py:53]: done

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+    # Check if directory with tensorboard logs exists
+    assert tensorboard_dir.exists(), "TensorBoard logs folder does not exist."
+    # Recursively search for files with tensorboard logger
+    event_files = list(tensorboard_dir.rglob("events.out.tfevents*"))
+    assert event_files, f"No TensorBoard event files found under {tensorboard_dir}"
+    assert len(matching_subfolders) == 1, "Only one checkpoint subfolder should be found."
+    if with_peft:
+        result_dir = tmp_path / "lora_finetune"
+        additional_args = "--lora-finetune " + mock_dataset_args
```
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py]: what would be the checkpoint for lora transform?

Comment (dorotat-nv) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py]: maybe additional_args are not needed since they are added in line 180

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py (hunk 1 above)
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py:33]: why do I need import `small_training_finetune_cmd`

File: sub-packages/bionemo-evo2/src/bionemo/evo2/models/peft.py
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/models/peft.py:1]: imho, this file should be under evo2/models or sth like this. could you relocate it? It was located here by mistake

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 7]
```diff
@@ -89,6 +88,27 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
         help="Use synthetic mock data for quick testing/debugging. Mutually exclusive with --dataset-config, --fasta-data and --sharded-eden-data.",
     )
 
+    parser.add_argument(
```
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: why do we need those arg? cant we control it by num_train_sample, num_val_samples and num_test_sample?

Comment (dorotat-nv) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: specific to dataset creation in general?

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py [hunk 8]
```diff
@@ -50,28 +50,6 @@ def run_train_with_std_redirect(args: argparse.Namespace) -> Tuple[str, nl.Train
...
-        f"train_evo2 --mock-data --result-dir {path} --devices {devices} "
-        "--model-size 1b_nv --num-layers 4 --hybrid-override-pattern SDH* --limit-val-batches 1 "
-        "--no-activation-checkpointing --add-bias-output --create-tensorboard-logger --create-tflops-callback "
-        f"--max-steps {max_steps} --warmup-steps 1 --val-check-interval {val_check} --limit-val-batches 1 "
-        f"--seq-length 16 --hidden-dropout 0.1 --attention-dropout 0.1 {additional_args}"
-    )
-    return cmd
-
-
-def small_training_finetune_cmd(path, max_steps, val_check, prev_ckpt, devices: int = 1, additional_args: str = ""):
```
Comment (dorotat-nv) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py:64]: what is the difference between the unittest test_train_evo2_finetune_runs in test_train and test_finetune? 
Ideally since we created test_finetune, would be great to have one place of tests related to finetuning

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py (hunk 7 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: The command didn't have options to specify the size of the dataset (number of samples) . I added those options so that they can be specified. The defaults create a dataset that is too big and it would take a long time to show the decrease in val loss with such a big dataset. In the test I specify to create a much smaller dataset.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py (hunk 1 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py:33]: Because the original test in `test_train` also tested finetune (with out LoRA) Even though I am adding a specialized tests on finetuning I still wanted to leave the original tests as is.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py (hunk 8 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_train.py:64]: the main difference is that the new tests adds PEFT case. Yes, we can in the future consolidate it the new file.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py (hunk 6 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py]: The `additional_args` set here are used in line 128

Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py]: At this point of the test we are doing LoRA finetuning for the first time. The LoRA checkpoint will be saved in `tmp_path / "lora_finetune"`

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+import tensorflow as tf
+from tensorflow.python.summary.summary_iterator import summary_iterator
+
```
[the remark this answers is not in the document]
Comment (balvisio) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py:22]: outdated. not using tf.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py (hunk 6 above)
Comment (dorotat-nv) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/run/test_finetune.py]: aahh ok, so when with LoRa we provide checkpoint, it resumes transformation, otherwise it starts from scratches? Super, great to know!

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py (hunk 7 above)
Comment (dorotat-nv) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: ok

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 10]
```diff
@@ -669,6 +689,9 @@ def train(args: argparse.Namespace) -> nl.Trainer:
             seq_length=args.seq_length,
             micro_batch_size=args.micro_batch_size,
             global_batch_size=global_batch_size,
+            num_train_samples=args.mock_num_train_samples,
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: why not use `--max-steps` for this? That to me feels like the most 1-1 with other dataloaders. Similarly you can use num_val_steps for num_val_samples.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 11]
```diff
@@ -89,6 +88,27 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
         help="Use synthetic mock data for quick testing/debugging. Mutually exclusive with --dataset-config, --fasta-data and --sharded-eden-data.",
     )
 
+    parser.add_argument(
+        "--mock-num-train-samples",
+        type=int,
+        default=10_000_000,
+        help="Number of training samples to generate",
+    )
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: See comment below on re-using variables, like `--max-steps` or `--num-val-steps` wherever possible. That way you don't need to do much more than add `--mock-data` as another flag.

Comment (balvisio) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: done

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py (hunk 10 above)
Comment (balvisio) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: done
